### PR TITLE
added authorization token support

### PR DESCRIPTION
**Pull request by mkalkbrenner**

### CODE REVIEW

**Comment by mkalkbrenner** [comment]
@thomascorthals It would be great if you review that PR. If you agree on the implementation, I'll add some documentation about authorization methods.

BTW unfortunately set Authentication() was introduced as function name for basic authentication. I suggest not to change that and to explicitly name the others that will be added. See #831

File: src/Core/Client/Adapter/Curl.php [hunk 1]
```diff
@@ -107,6 +107,14 @@ public function createHandle(Request $request, Endpoint $endpoint)
             curl_setopt($handler, CURLOPT_USERPWD, $authData['username'].':'.$authData['password']);
             curl_setopt($handler, CURLOPT_HTTPAUTH, CURLAUTH_BASIC);
         }
+        elseif (!isset($options['headers']) || !array_key_exists('Authorization', $options['headers'])) {
```
Comment (thomascorthals) [inline src/Core/Client/Adapter/Curl.php:110]:
```suggestion
        elseif (!isset($options['headers']) || !\array_key_exists('Authorization', $options['headers'])) {
```

In keeping with https://github.com/FriendsOfPHP/PHP-CS-Fixer/blob/866715801ac121a0a4cde6fdb9d9a3dd6873b606/src/Fixer/FunctionNotation/NativeFunctionInvocationFixer.php#L347

File: src/Core/Client/Adapter/Curl.php [hunk 2]
```diff
@@ -107,6 +107,14 @@ public function createHandle(Request $request, Endpoint $endpoint)
...
+            // According to the specification, only one Authorization header is allowed.
+            // @se https://stackoverflow.com/questions/29282578/multiple-http-authorization-headers
```
Comment (thomascorthals) [inline src/Core/Client/Adapter/Curl.php:112]:
```suggestion
            // @see https://stackoverflow.com/questions/29282578/multiple-http-authorization-headers
```

File: src/Core/Client/Adapter/Http.php [hunk 3]
```diff
@@ -103,6 +103,16 @@ public function createContext(Request $request, Endpoint $endpoint)
                 'Authorization: Basic '.base64_encode($authData['username'].':'.$authData['password'])
             );
         }
+        else {
+            // According to the specification, only one Authorization header is allowed.
+            // @se https://stackoverflow.com/questions/29282578/multiple-http-authorization-headers
```
Comment (thomascorthals) [inline src/Core/Client/Adapter/Http.php:108]:
```suggestion
            // @see https://stackoverflow.com/questions/29282578/multiple-http-authorization-headers
```

File: src/Core/Client/Adapter/Psr18Adapter.php [hunk 4]
```diff
@@ -171,6 +171,14 @@ private function getRequestHeaders(Request $request, Endpoint $endpoint): array
             if (!empty($authData['username']) && !empty($authData['password'])) {
                 $headers['Authorization'] = ['Basic '.base64_encode($authData['username'].':'.$authData['password'])];
             }
+            else {
+                // According to the specification, only one Authorization header is allowed.
+                // @se https://stackoverflow.com/questions/29282578/multiple-http-authorization-headers
```
Comment (thomascorthals) [inline src/Core/Client/Adapter/Psr18Adapter.php:176]:
```suggestion
                // @see https://stackoverflow.com/questions/29282578/multiple-http-authorization-headers
```

**Review by thomascorthals — CHANGES_REQUESTED** [review]
Looks good. Just some nitpicks on top of the ones from StyleCI.

And it appears you didn't commit an upated `CurlTest`?

**Comment by mkalkbrenner** [comment]
@thomascorthals we don't have any authorization tests for curl so far. So we should also add something for basic auth.

**Review by thomascorthals — DISMISSED** [review]

**Comment by thomascorthals** [comment]
I assume I'm not the only one getting annoyed that there is no `curl_getopt()` for testing stuff like this.

**Pull request merged**